File: frontend/mobile/lib/core/services/recording_service.dart
import 'dart:async';
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter_sound/flutter_sound.dart';
import 'package:path_provider/path_provider.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:uuid/uuid.dart';
import '../features/recitation/data/models/recording_models.dart';

/// Service for audio recording with waveform visualization
class RecordingService {
  final FlutterSoundRecorder _recorder = FlutterSoundRecorder();
  final StreamController<WaveformData> _waveformController =
      StreamController<WaveformData>.broadcast();
  final StreamController<RecordingState> _stateController =
      StreamController<RecordingState>.broadcast();
  final StreamController<Duration> _durationController =
      StreamController<Duration>.broadcast();

  RecordingState _currentState = RecordingState.idle;
  String? _currentRecordingPath;
  DateTime? _recordingStartTime;
  Timer? _durationTimer;
  final List<double> _amplitudes = [];
  AudioQuality _currentQuality = AudioQuality.high;

  // Getters
  Stream<WaveformData> get waveformStream => _waveformController.stream;
  Stream<RecordingState> get stateStream => _stateController.stream;
  Stream<Duration> get durationStream => _durationController.stream;
  RecordingState get currentState => _currentState;
  String? get currentRecordingPath => _currentRecordingPath;

  /// Initialize the recording service
  Future<void> initialize() async {
    await _recorder.openRecorder();
    await _recorder.setSubscriptionDuration(const Duration(milliseconds: 100));
  }

  /// Check and request microphone permission
  Future<bool> checkPermission() async {
    final status = await Permission.microphone.status;
    if (status.isGranted) {
      return true;
    }

    final result = await Permission.microphone.request();
    return result.isGranted;
  }

  /// Start recording with specified quality
  Future<void> startRecording({
    required int surahNumber,
    required int ayahStart,
    required int ayahEnd,
    AudioQuality quality = AudioQuality.high,
  }) async {
    if (_currentState == RecordingState.recording) {
      throw Exception('Already recording');
    }

    // Check permission
    final hasPermission = await checkPermission();
    if (!hasPermission) {
      _updateState(RecordingState.error);
      throw Exception('Microphone permission not granted');
    }

    _updateState(RecordingState.preparing);
    _currentQuality = quality;
    _amplitudes.clear();

    // Generate file path
    final directory = await getApplicationDocumentsDirectory();
    final recordingId = const Uuid().v4();
    _currentRecordingPath =
        '${directory.path}/recordings/$recordingId.wav';

    // Create directory if it doesn't exist
    final recordingDir = Directory('${directory.path}/recordings');
    if (!await recordingDir.exists()) {
      await recordingDir.create(recursive: true);
    }

    // Start recording
    await _recorder.startRecorder(
      toFile: _currentRecordingPath,
      codec: Codec.pcm16WAV,
      sampleRate: quality.sampleRate,
      numChannels: 1,
    );

    _recordingStartTime = DateTime.now();
    _updateState(RecordingState.recording);

    // Start duration timer
    _startDurationTimer();

    // Listen to amplitude for waveform
    _recorder.onProgress!.listen((event) {
      if (event.decibels != null) {
        final amplitude = _decibelToAmplitude(event.decibels!);
        _amplitudes.add(amplitude);

        // Update waveform every 100ms
        if (_amplitudes.length % 10 == 0) {
          _updateWaveform();
        }
      }
    });
  }

  /// Pause recording
  Future<void> pauseRecording() async {
    if (_currentState != RecordingState.recording) {
      return;
    }

    await _recorder.pauseRecorder();
    _updateState(RecordingState.paused);
    _durationTimer?.cancel();
  }

  /// Resume recording
  Future<void> resumeRecording() async {
    if (_currentState != RecordingState.paused) {
      return;
    }

    await _recorder.resumeRecorder();
    _updateState(RecordingState.recording);
    _startDurationTimer();
  }

  /// Stop recording and return metadata
  Future<RecordingMetadata> stopRecording({
    required int surahNumber,
    required int ayahStart,
    required int ayahEnd,
  }) async {
    if (_currentState != RecordingState.recording &&
        _currentState != RecordingState.paused) {
      throw Exception('Not recording');
    }

    _updateState(RecordingState.processing);
    _durationTimer?.cancel();

    await _recorder.stopRecorder();

    final duration = DateTime.now().difference(_recordingStartTime!);
    final file = File(_currentRecordingPath!);
    final fileSize = await file.length();

    final metadata = RecordingMetadata(
      recordingId: const Uuid().v4(),
      surahNumber: surahNumber,
      ayahStart: ayahStart,
      ayahEnd: ayahEnd,
      quality: _currentQuality,
      recordedAt: _recordingStartTime!,
      duration: duration,
      fileSizeBytes: fileSize,
      filePath: _currentRecordingPath,
    );

    _updateState(RecordingState.stopped);
    _currentRecordingPath = null;
    _recordingStartTime = null;

    return metadata;
  }

  /// Cancel recording and delete file
  Future<void> cancelRecording() async {
    if (_currentState == RecordingState.recording ||
        _currentState == RecordingState.paused) {
      await _recorder.stopRecorder();
      _durationTimer?.cancel();

      // Delete the recording file
      if (_currentRecordingPath != null) {
        final file = File(_currentRecordingPath!);
        if (await file.exists()) {
          await file.delete();
        }
      }

      _currentRecordingPath = null;
      _recordingStartTime = null;
      _amplitudes.clear();
      _updateState(RecordingState.idle);
    }
  }

  /// Get current recording duration
  Duration getCurrentDuration() {
    if (_recordingStartTime == null) {
      return Duration.zero;
    }
    return DateTime.now().difference(_recordingStartTime!);
  }

  /// Convert decibels to normalized amplitude (0.0 to 1.0)
  double _decibelToAmplitude(double decibels) {
    // Normalize decibels (-160 to 0) to amplitude (0.0 to 1.0)
    const minDb = -60.0;
    const maxDb = 0.0;
    final normalized = (decibels - minDb) / (maxDb - minDb);
    return normalized.clamp(0.0, 1.0);
  }

  /// Update waveform data
  void _updateWaveform() {
    if (_amplitudes.isEmpty) return;

    final maxAmplitude = _amplitudes.reduce((a, b) => a > b ? a : b);
    final minAmplitude = _amplitudes.reduce((a, b) => a < b ? a : b);

    final waveformData = WaveformData(
      amplitudes: List.from(_amplitudes),
      maxAmplitude: maxAmplitude,
      minAmplitude: minAmplitude,
      duration: getCurrentDuration(),
    );

    _waveformController.add(waveformData);
  }

  /// Update recording state
  void _updateState(RecordingState state) {
    _currentState = state;
    _stateController.add(state);
  }

  /// Start duration timer
  void _startDurationTimer() {
    _durationTimer?.cancel();
    _durationTimer = Timer.periodic(const Duration(seconds: 1), (timer) {
      _durationController.add(getCurrentDuration());
    });
  }

  /// Dispose resources
  Future<void> dispose() async {
    _durationTimer?.cancel();
    await _recorder.closeRecorder();
    await _waveformController.close();
    await _stateController.close();
    await _durationController.close();
  }
}

File: frontend/nextjs-app/src/lib/services/recording-service.ts
import {
  AudioQuality,
  AudioQualitySettings,
  RecordingState,
  WaveformData,
  RecordingMetadata,
} from '@/types/recording';

export class RecordingService {
  private mediaRecorder: MediaRecorder | null = null;
  private audioContext: AudioContext | null = null;
  private analyser: AnalyserNode | null = null;
  private dataArray: Uint8Array | null = null;
  private stream: MediaStream | null = null;
  private chunks: Blob[] = [];
  private startTime: number = 0;
  private animationFrameId: number | null = null;

  private stateListeners: Set<(state: RecordingState)> = new Set();
  private waveformListeners: Set<(data: WaveformData)> = new Set();
  private durationListeners: Set<(duration: number)> = new Set();

  private currentState: RecordingState = RecordingState.IDLE;
  private amplitudes: number[] = [];

  constructor() {
    if (typeof window !== 'undefined') {
      this.audioContext = new (window.AudioContext || (window as any).webkitAudioContext)();
    }
  }

  // Event listeners
  onStateChange(listener: (state: RecordingState) => void): () => void {
    this.stateListeners.add(listener);
    return () => this.stateListeners.delete(listener);
  }

  onWaveformUpdate(listener: (data: WaveformData) => void): () => void {
    this.waveformListeners.add(listener);
    return () => this.waveformListeners.delete(listener);
  }

  onDurationUpdate(listener: (duration: number) => void): () => void {
    this.durationListeners.add(listener);
    return () => this.durationListeners.delete(listener);
  }

  private updateState(state: RecordingState): void {
    this.currentState = state;
    this.stateListeners.forEach(listener => listener(state));
  }

  private updateWaveform(): void {
    if (!this.analyser || !this.dataArray) return;

    this.analyser.getByteTimeDomainData(this.dataArray);

    // Convert to normalized amplitudes
    const amplitudes = Array.from(this.dataArray).map(
      value => Math.abs((value - 128) / 128)
    );

    // Store amplitudes
    this.amplitudes.push(...amplitudes);

    // Keep only recent amplitudes (last 1000 samples)
    if (this.amplitudes.length > 1000) {
      this.amplitudes = this.amplitudes.slice(-1000);
    }

    const maxAmplitude = Math.max(...this.amplitudes);
    const minAmplitude = Math.min(...this.amplitudes);
    const duration = Date.now() - this.startTime;

    const waveformData: WaveformData = {
      amplitudes: this.amplitudes,
      maxAmplitude,
      minAmplitude,
      duration,
    };

    this.waveformListeners.forEach(listener => listener(waveformData));
  }

  private updateDuration(): void {
    if (this.currentState === RecordingState.RECORDING) {
      const duration = Date.now() - this.startTime;
      this.durationListeners.forEach(listener => listener(duration));
    }
  }

  private startWaveformAnalysis(): void {
    const analyze = () => {
      if (this.currentState === RecordingState.RECORDING) {
        this.updateWaveform();
        this.updateDuration();
        this.animationFrameId = requestAnimationFrame(analyze);
      }
    };
    analyze();
  }

  private stopWaveformAnalysis(): void {
    if (this.animationFrameId !== null) {
      cancelAnimationFrame(this.animationFrameId);
      this.animationFrameId = null;
    }
  }

  async checkPermission(): Promise<boolean> {
    try {
      const result = await navigator.permissions.query({ name: 'microphone' as PermissionName });
      return result.state === 'granted';
    } catch {
      // Fallback: try to get user media
      try {
        const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
        stream.getTracks().forEach(track => track.stop());
        return true;
      } catch {
        return false;
      }
    }
  }

  async startRecording(quality: AudioQuality = AudioQuality.HIGH): Promise<void> {
    if (this.currentState === RecordingState.RECORDING) {
      throw new Error('Already recording');
    }

    this.updateState(RecordingState.PREPARING);

    try {
      // Get user media
      this.stream = await navigator.mediaDevices.getUserMedia({
        audio: {
          sampleRate: AudioQualitySettings[quality].sampleRate,
          channelCount: 1,
          echoCancellation: true,
          noiseSuppression: true,
          autoGainControl: true,
        },
      });

      // Setup audio context and analyser
      if (!this.audioContext) {
        this.audioContext = new AudioContext();
      }

      const source = this.audioContext.createMediaStreamSource(this.stream);
      this.analyser = this.audioContext.createAnalyser();
      this.analyser.fftSize = 2048;
      this.analyser.smoothingTimeConstant = 0.8;

      source.connect(this.analyser);

      const bufferLength = this.analyser.frequencyBinCount;
      this.dataArray = new Uint8Array(bufferLength);

      // Setup media recorder
      const mimeType = this.getSupportedMimeType();
      this.mediaRecorder = new MediaRecorder(this.stream, {
        mimeType,
        audioBitsPerSecond: AudioQualitySettings[quality].bitRate,
      });

      this.chunks = [];
      this.amplitudes = [];

      this.mediaRecorder.ondataavailable = (event) => {
        if (event.data.size > 0) {
          this.chunks.push(event.data);
        }
      };

      this.mediaRecorder.onstop = () => {
        this.stopWaveformAnalysis();
      };

      // Start recording
      this.mediaRecorder.start(100); // Collect data every 100ms
      this.startTime = Date.now();
      this.updateState(RecordingState.RECORDING);

      // Start waveform analysis
      this.startWaveformAnalysis();
    } catch (error) {
      this.updateState(RecordingState.ERROR);
      throw error;
    }
  }

  async pauseRecording(): Promise<void> {
    if (this.currentState !== RecordingState.RECORDING || !this.mediaRecorder) {
      return;
    }

    this.mediaRecorder.pause();
    this.stopWaveformAnalysis();
    this.updateState(RecordingState.PAUSED);
  }

  async resumeRecording(): Promise<void> {
    if (this.currentState !== RecordingState.PAUSED || !this.mediaRecorder) {
      return;
    }

    this.mediaRecorder.resume();
    this.startWaveformAnalysis();
    this.updateState(RecordingState.RECORDING);
  }

  async stopRecording(metadata: {
    surahNumber: number;
    ayahStart: number;
    ayahEnd: number;
    quality: AudioQuality;
  }): Promise<RecordingMetadata> {
    if (
      (this.currentState !== RecordingState.RECORDING &&
        this.currentState !== RecordingState.PAUSED) ||
      !this.mediaRecorder
    ) {
      throw new Error('Not recording');
    }

    this.updateState(RecordingState.PROCESSING);

    return new Promise((resolve, reject) => {
      if (!this.mediaRecorder) {
        reject(new Error('MediaRecorder not initialized'));
        return;
      }

      this.mediaRecorder.onstop = () => {
        const blob = new Blob(this.chunks, { type: this.getSupportedMimeType() });
        const duration = Date.now() - this.startTime;

        const recordingMetadata: RecordingMetadata = {
          recordingId: this.generateId(),
          surahNumber: metadata.surahNumber,
          ayahStart: metadata.ayahStart,
          ayahEnd: metadata.ayahEnd,
          quality: metadata.quality,
          recordedAt: new Date(),
          duration,
          fileSizeBytes: blob.size,
          blob,
        };

        this.cleanup();
        this.updateState(RecordingState.STOPPED);
        resolve(recordingMetadata);
      };

      this.mediaRecorder.stop();
    });
  }

  async cancelRecording(): Promise<void> {
    if (
      this.currentState === RecordingState.RECORDING ||
      this.currentState === RecordingState.PAUSED
    ) {
      if (this.mediaRecorder && this.mediaRecorder.state !== 'inactive') {
        this.mediaRecorder.stop();
      }
      this.cleanup();
      this.updateState(RecordingState.IDLE);
    }
  }

  private cleanup(): void {
    this.stopWaveformAnalysis();

    if (this.stream) {
      this.stream.getTracks().forEach(track => track.stop());
      this.stream = null;
    }

    this.mediaRecorder = null;
    this.analyser = null;
    this.dataArray = null;
    this.chunks = [];
    this.amplitudes = [];
  }

  private getSupportedMimeType(): string {
    const types = [
      'audio/webm;codecs=opus',
      'audio/webm',
      'audio/ogg;codecs=opus',
      'audio/ogg',
      'audio/wav',
    ];

    for (const type of types) {
      if (MediaRecorder.isTypeSupported(type)) {
        return type;
      }
    }

    return 'audio/webm'; // Fallback
  }

  private generateId(): string {
    return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  dispose(): void {
    this.cleanup();
    if (this.audioContext) {
      this.audioContext.close();
      this.audioContext = null;
    }
    this.stateListeners.clear();
    this.waveformListeners.clear();
    this.durationListeners.clear();
  }
}
